csrc/core/scalar_type.hpp

```diff
@@ -79,19 +79,74 @@ class ScalarType {
   NanRepr const nan_repr;         // how NaNs are represented
                                   // (not applicable for integer types)
 
-  int64_t size_bits() const { return mantissa + exponent + is_signed(); }
-  bool is_signed() const { return signed_; }
-  bool is_integer() const { return exponent == 0; }
-  bool is_floating_point() const { return exponent > 0; }
-  bool is_ieee_754() const {
+  // unique id for this scalar type that can be computed at compile time for
+  //  c++17 tempalate specialization this is not needed once we migrate to c++20
+  //  and can pass literal classes as template parameters
+  constexpr int64_t id() const {
+    int64_t id = 0;
+    uint32_t bit_offset = 0;
+
+    auto or_and_advance = [&](auto val) {
+      auto constexpr val_size_bits =
+          std::is_same_v<decltype(val), bool> ? 1 : sizeof(val) * 8;
+      id |= (int64_t(val) & ((uint64_t(1) << val_size_bits) - 1)) << bit_offset;
+      bit_offset += val_size_bits;
+    };
+
+    or_and_advance(exponent);
+    or_and_advance(mantissa);
+    or_and_advance(bias);
+    or_and_advance(signed_);
+    or_and_advance(finite_values_only);
+    or_and_advance(nan_repr);
+    return id;
+  }
```

Comment:
Could you add a `static_assert` to make sure we don't exceed the size of an `int64_t`, and a warning comment as well?

I suggest `static_assert(sizeof(ScalarType) < sizeof(id)`, assuming that this function needs to take all member variables of ScalarType into account

Reply:
The challenge here is the `bool`s, in the id we use a single bit for these but `static_assert(sizeof(ScalarType) < sizeof(id)` would not take this into account unfortunately :(

Reply:
good point -- go with @bnellnm's suggestion :)

Reply:
haha I was I could: see https://github.com/vllm-project/vllm/pull/7323#discussion_r1712008722

stuff will get so much better with C++20 ...